src/hotspot/share/opto/vectorization.hpp

```diff
@@ -1323,6 +1323,350 @@ class AlignmentSolver {
+  void trace_schedule_cycle(const GrowableArray<VTransformNode*>& stack,
+                            const VectorSet& pre_visited,
+                            const VectorSet& post_visited) const;
+#endif
+};
+
+// VTransformNodes resemble the C2 IR Nodes. They represent the resulting scalar and
+// vector nodes as closely as possible.
+class VTransformNode : public ArenaObj {
+public:
+  const VTransformNodeIDX _idx;
+
+private:
+  // _in is split into required inputs (_req), and additional dependencies.
+  const uint _req;
+  GrowableArray<VTransformNode*> _in;
+  GrowableArray<VTransformNode*> _out;
+
+public:
+  VTransformNode(VTransformGraph& graph, const uint req) :
+    _idx(graph.new_idx()),
+    _req(req),
+    _in(graph.arena(),  req, req, nullptr),
+    _out(graph.arena(), 4, 0, nullptr)
+  {
+    graph.add_vtnode(this);
+  }
+
+  void set_req(uint i, VTransformNode* n) {
+    assert(i < _req, "must be a req");
+    assert(_in.at(i) == nullptr && n != nullptr, "only set once");
+    _in.at_put(i, n);
+    n->add_out(this);
+  }
+
+  void swap_req(uint i, uint j) {
+    assert(i < _req, "must be a req");
+    assert(j < _req, "must be a req");
+    VTransformNode* tmp = _in.at(i);
+    _in.at_put(i, _in.at(j));
+    _in.at_put(j, tmp);
+  }
+
+  void add_dependency(VTransformNode* n) {
+    assert(n != nullptr, "no need to add nullptr");
+    _in.push(n);
+    n->add_out(this);
+  }
+
+  void add_out(VTransformNode* n) {
+    _out.push(n);
+  }
+
+  uint req() const { return _req; }
+  VTransformNode* in(int i) const { return _in.at(i); }
+  int outs() const { return _out.length(); }
+  VTransformNode* out(int i) const { return _out.at(i); }
+
+  bool has_req_or_dep() const {
+    for (int i = 0; i < _in.length(); i++) {
+      if (_in.at(i) != nullptr) { return true; }
+    }
+    return false;
+  }
+
+  virtual VTransformScalarNode* isa_Scalar() { return nullptr; }
+  virtual VTransformInputScalarNode* isa_InputScalar() { return nullptr; }
+  virtual VTransformVectorNode* isa_Vector() { return nullptr; }
+  virtual VTransformElementWiseVectorNode* isa_ElementWiseVector() { return nullptr; }
+  virtual VTransformBoolVectorNode* isa_BoolVector() { return nullptr; }
+  virtual VTransformReductionVectorNode* isa_ReductionVector() { return nullptr; }
+
+  virtual VTransformApplyStatus apply(const VLoopAnalyzer& vloop_analyzer,
+                                      const GrowableArray<Node*>& vnode_idx_to_transformed_node) const = 0;
+
+  Node* find_transformed_input(int i, const GrowableArray<Node*>& vnode_idx_to_transformed_node) const;
+
+  void register_new_node_from_vectorization(const VLoopAnalyzer& vloop_analyzer, Node* vn, Node* old_node) const;
+
+  NOT_PRODUCT(virtual const char* name() const = 0;)
+  NOT_PRODUCT(void print() const;)
+  NOT_PRODUCT(virtual void print_spec() const {};)
+  NOT_PRODUCT(static void print_node_idx(const VTransformNode* vtn);)
+};
+
+class VTransformScalarNode : public VTransformNode {
```

Comment:
Not sure if you left out comments above the vtnode classes on purpose. But maybe a short summary about them could help readers of the code (if the class itself is not already self-explanatory).

Reply:
Ok, will add some for each class.